File: frontend/src/app/services/carrito.service.ts
import { Injectable, computed, signal } from '@angular/core';

export interface CarritoItem {
  variante_id: number;
  cantidad: number;
  nombre: string;
  sku: string;
  color?: string | null;
  talla?: string | null;
  precio: number;
  imagen_url?: string | null;
}

const CLAVE = 'fashionstore_carrito_reservas';

@Injectable({ providedIn: 'root' })
export class CarritoService {
  private itemsSig = signal<CarritoItem[]>(this.leer());

  readonly items = this.itemsSig.asReadonly();
  readonly totalPrendas = computed(() =>
    this.items().reduce((acc, i) => acc + i.cantidad, 0),
  );
  readonly vacio = computed(() => this.items().length === 0);

  agregar(item: CarritoItem): void {
    const lista = [...this.itemsSig()];
    const idx = lista.findIndex((i) => i.variante_id === item.variante_id);
    if (idx >= 0) {
      lista[idx] = {
        ...lista[idx],
        cantidad: lista[idx].cantidad + item.cantidad,
      };
    } else {
      lista.push(item);
    }
    this.itemsSig.set(lista);
    this.guardar();
  }

  cambiarCantidad(varianteId: number, cantidad: number): void {
    this.itemsSig.set(
      this.itemsSig().map((i) =>
        i.variante_id === varianteId
          ? { ...i, cantidad: Math.max(1, cantidad) }
          : i,
      ),
    );
    this.guardar();
  }

  quitar(varianteId: number): void {
    this.itemsSig.set(this.itemsSig().filter((i) => i.variante_id !== varianteId));
    this.guardar();
  }

  limpiar(): void {
    this.itemsSig.set([]);
    this.guardar();
  }

  private leer(): CarritoItem[] {
    try {
      const raw = localStorage.getItem(CLAVE);
      return raw ? (JSON.parse(raw) as CarritoItem[]) : [];
    } catch {
      return [];
    }
  }

  private guardar(): void {
    localStorage.setItem(CLAVE, JSON.stringify(this.itemsSig()));
  }
}

File: mobile/lib/services/carrito_service.dart
import 'dart:convert';

import 'package:shared_preferences/shared_preferences.dart';

import '../models/carrito.dart';

class CarritoService {
  CarritoService._();

  static final CarritoService instance = CarritoService._();

  static const _clave = 'fashionstore_carrito_reservas';

  final List<CarritoItem> _items = [];
  bool _iniciado = false;

  List<CarritoItem> get items => List.unmodifiable(_items);
  int get totalPrendas => _items.fold(0, (acc, item) => acc + item.cantidad);
  bool get vacio => _items.isEmpty;

  Future<void> init() async {
    if (_iniciado) return;
    _iniciado = true;
    try {
      final prefs = await SharedPreferences.getInstance();
      final raw = prefs.getString(_clave);
      if (raw == null || raw.isEmpty) return;
      final lista = (jsonDecode(raw) as List).cast<Map<String, dynamic>>();
      _items.addAll(lista.map(CarritoItem.fromJson));
    } catch (_) {
      _items.clear();
    }
  }

  Future<void> agregar(CarritoItem item) async {
    await init();
    final idx = _items.indexWhere((i) => i.varianteId == item.varianteId);
    if (idx >= 0) {
      _items[idx].cantidad += item.cantidad;
    } else {
      _items.add(item);
    }
    await _guardar();
  }

  Future<void> cambiarCantidad(int varianteId, int cantidad) async {
    final idx = _items.indexWhere((i) => i.varianteId == varianteId);
    if (idx >= 0) {
      _items[idx].cantidad = cantidad < 1 ? 1 : cantidad;
      await _guardar();
    }
  }

  Future<void> quitar(int varianteId) async {
    _items.removeWhere((i) => i.varianteId == varianteId);
    await _guardar();
  }

  Future<void> limpiar() async {
    _items.clear();
    await _guardar();
  }

  Future<void> _guardar() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(
      _clave,
      jsonEncode(_items.map((i) => i.toJson()).toList()),
    );
  }
}